packages/esm-ohri-pmtct-app/src/form-entry/post-submission-actions/mother-child-linkange-action.ts

```diff
@@ -0,0 +1,114 @@
+const MotherToChildRelationshipType = '8d91a210-c2cc-11de-8d13-0010c6dffd0f';
+
+export const MotherToChildLinkageSubmissionAction: PostSubmissionAction = {
+  applyAction: async function ({ patient, encounters, sessionMode }) {
+    const encounter = encounters[0];
+    const encounterLocation = encounter.location['uuid'];
+    // only do this the first time the form is entered
+    if (sessionMode !== 'enter') {
+      return;
+    }
+    const infantsToCreate = await Promise.all(
+      findObsByConcept(encounter, infantDetailsGroup).map(async (obsGroup) =>
+        constructPatientObjectFromObsData(obsGroup, encounterLocation),
+      ),
+    );
+    const postResponse = await savePatients(infantsToCreate);
+    postResponse.map(({ data }) =>
+      saveRelationship({
+        relationshipType: MotherToChildRelationshipType,
+        personA: patient.id,
+        personB: data.uuid,
+      }),
+    );
+  },
+};
+
+async function constructPatientObjectFromObsData(obsGroup, encounterLocation: string): Promise<Patient> {
+  // check if infant is alive
+  const lifeStatusAtBirth = findChildObsInTree(obsGroup, infantLifeStatus);
+  if (getObsValueCoded(lifeStatusAtBirth) == aliveStatus) {
+    // the infant is alive hence eligible for registration
+    const patient: Patient = {
+      identifiers: [],
+      person: {
+        names: [
+          {
+            givenName: 'TBD',
+            middleName: 'TBD',
+            familyName: 'TBD',
+            preferred: true,
+          },
+        ],
+        gender: inferGenderFromObs(findChildObsInTree(obsGroup, infantGender)),
+        birthdate: findChildObsInTree(obsGroup, infantDOB)?.value,
+        birthdateEstimated: false,
+        dead: false,
+        deathDate: null,
+        causeOfDeath: '',
+      },
+    };
+    // PTracker ID
+    const pTrackerId = findChildObsInTree(obsGroup, infantPTrackerId)?.value;
+    if (pTrackerId) {
+      patient.identifiers = [
+        {
+          identifier: pTrackerId,
+          location: encounterLocation,
+          identifierType: PtrackerIdentifierType,
+          preferred: false,
+        },
+      ];
+    }
+    // generate the preferred identifier
+    const preferredIdentifier: PatientIdentifier = {
+      identifier: await (await generateIdentifier(preferredIdentifierSource)).data.identifier,
+      identifierType: OpenmrsClassicIdentifierType,
+      location: encounterLocation,
+      preferred: true,
+    };
+    patient.identifiers.push(preferredIdentifier);
+    return patient;
+  }
+  return null;
+}
+
+////////////////////////
+// Convinience functions
+////////////////////////
+
+function findObsByConcept(encounter: any, concept: string): Array<any> {
+  return encounter?.obs?.filter((observation) => observation.concept.uuid === concept) || [];
+}
+
+function findChildObsInTree(parent: any, childConcept: string) {
+  return parent.groupMembers?.find((obs) => obs.concept.uuid == childConcept);
+}
+
```

Comment:
are these functions that could be reused in other components? is so wouldn't it be better to move them to a separate file or the commons?

Reply:
Sure thing @pirupius, I think putting them in a separate file would be a good starting point until we find a compelling usecase to move them to commons.